File: 0015-3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        dic = {}
        res = []
        for i, x in enumerate(nums):
            l = i + 1
            r = len(nums) - 1
            if i > 0 and x == nums[i - 1]:
                continue
            while l < r:
                if nums[l] + nums[r] + x == 0:
                    res.append(sorted([nums[l], nums[r], x]))
                    l += 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
                elif nums[l] + nums[r] + x > 0:
                    r -= 1
                else:
                    l += 1
        return res

File: 0015-3sum/solution.cpp
using namespace std;
#include <vector>
#include <set>
#include <algorithm>
#include <iostream>
static const int speedup = []() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);
    return 0;
}();

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res = {};
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); ++i){
            if (i > 0 and (nums[i] == nums[i - 1])) continue;
            int l = i + 1;
            int r = nums.size() - 1;
            while (l < r){
                int sum = nums[l] + nums[r] + nums[i];
                if (sum > 0){
                    r--;
                }
                else if(sum < 0){
                    l++;
                }
                else{
                    vector<int> temp = {nums[l], nums[r], nums[i]};
                    res.push_back(temp);
                    while(l < r and nums[l + 1] == nums[l]) l++;
                    while(l < r and nums[r - 1] == nums[r]) r--;
                    r--;
                    l++;
                }
            }
        }
        return res;
        
    }
};
